assets/app/features/upgrade/steps.factory.spec.js

```diff
@@ -118,5 +406,87 @@ describe('Stepes Factory', function () {
+                expect(upgradeStepsFactory.stepByID(9999)).not.toBeDefined();
+            });
+        });
+
+        describe('lastStateForRestore function', function () {
+            describe('when there is no stored last state', function () {
+                beforeEach(function () {
+                    localStorage.removeItem(UPGRADE_LAST_STATE_KEY);
+                });
+                it('should return "upgrade-landing" before any step', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(initialStatusData))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade-landing" after prechecks', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterPrechecks))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade-landing" during prepare', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataDuringPrepare))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade.backup" after prepare', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterPrepare))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.backup" during backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataDuringBackup))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.backup" after backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterBackup))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.administration-repository-checks" after admin repo checks', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterAdminRepoChecks))
+                        .toEqual('upgrade.administration-repository-checks');
+                });
+            });
+            describe('when stored state is invalid in given case', function () {
+                beforeEach(function () {
+                    localStorage.setItem(UPGRADE_LAST_STATE_KEY, 'upgrade-landing');
+                });
+                it('should return "upgrade.backup" after backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterBackup))
+                        .toEqual('upgrade.backup');
+                });
+            });
+            describe('when stored state points to current, pending step', function () {
+                beforeEach(function () {
+                    localStorage.setItem(UPGRADE_LAST_STATE_KEY, 'upgrade.administration-repository-checks');
```

Comment:
I added more validations to the section without localStorage part. These are supposed to ensure the mapping is correct.